### PR TITLE
[5.1] Add Pre Task hook to Scheduler

### PR DESCRIPTION
**Pull request by laurencei**
Hi. There is currently a Post Task hook for the scheduler (using `then()`)

It would be nice if you could hook into the Pre Task, to run anything required before the command itself runs.

A specific use case for this is to Ping a monitoring service that your task is _about_ to start - so you can track/monitor the length of time the task took to run.

If this PR is accepted I'll do up a modification for the Docs to show it there as well.

p.s. I used the naming convention of `before()` and `beforePing()` - but it might be before to call it `now()` and `nowPing()` which is kind of the opposite of `then()`? Let me know if you want me to change it...

### CODE REVIEW

File: src/Illuminate/Console/Scheduling/Event.php [hunk 1]
```diff
@@ -117,7 +124,7 @@ public function __construct($command)
      */
     public function run(Container $container)
     {
-        if (count($this->afterCallbacks) > 0) {
+        if ((count($this->afterCallbacks) > 0) || (count($this->beforeCallbacks) > 0)) {
```
Comment (GrahamCampbell) [inline src/Illuminate/Console/Scheduling/Event.php]: You don't need the extra brackets.

**Comment by GrahamCampbell** [comment]
:-1: You should be using middleware here. We already removed filters in the router. I don't think we should add stuff here too.

**Comment by TheShiftExchange** [comment]
@GrahamCampbell - I'm confused - what does this have to do with filters or middleware?

We are saying we want to run a specific task(s) *before* we run the command.

It is not a filter (i.e .to check if we *want* to run the command) - the command will always run after the `before()` command.

And how would you apply middleware here anyway?

**Comment by GrahamCampbell** [comment]
> We are saying we want to run a specific task(s) before we run the command.

Yeh, it's the same idea. 

> And how would you apply middleware here anyway? Its not a HTTP request.

Middleware don't have to apply to routing. The exist in the command system of laravel too.

**Comment by GrahamCampbell** [comment]
I don't want to add filters anywhere new. Laravel's already decided they're a bad design pattern and they're removed in 5.2.

**Comment by TheShiftExchange** [comment]
But these are not `filters`?

Then a pre-coomand hooks - to run additional commands prior to the command itself running?

It is no different to the Post hook function - that allows you to run additional commands after the command itself?

**Comment by GrahamCampbell** [comment]
> These are pre-command hooks - to run additional commands prior to the command itself running?

That's exactly what middleware are for.

**Comment by TheShiftExchange** [comment]
...so by that same logic - we should actually remove `then()` - since you can just use middleware to run tasks after the command is run?

**Comment by GrahamCampbell** [comment]
Yeh. Though we don't have middleware for this yet. I'm saying we should do that instead of more filter type stuff. Currently we only have middleware for the command bus.

**Comment by taylorotwell** [comment]
I'm fine with this other than the wording of `beforePing`. That reads awkwardly to me.

**Comment by TheShiftExchange** [comment]
@taylorotwell - what about 'first()' and 'firstPing()'

Or

'firstly()' and 'firstlyPing()'

**Comment by ibrasho** [comment]
```php
$schedule->command('inspire')
    ->firstPing('http://callback/endpoint')
    ->then(function () { doSomething(); });
```

```php
$schedule->command('inspire')
    ->pingBefore('http://callback/endpoint')
    ->then(function () { doSomething(); });
```

**Comment by TheShiftExchange** [comment]
@taylorotwell - I've changed it to `first()` and `firstPing()` - which I thinks reads the best? 

```
$schedule->command('inspire')
    ->firstPing('http://callback/endpoint')
    ->then(function () { doSomething(); });
```

If this is accepted I'll do a PR on the Docs for you...

**Pull request merged**